Tkt 00184 - Update header layout

Android63 opened the pull request:
Updated header from this:
![screen shot 2018-01-29 at 13 12 34](https://user-images.githubusercontent.com/1991226/35512016-28142780-04f6-11e8-839f-0c2d13506242.png)

To this:
![screen shot 2018-01-29 at 13 14 31](https://user-images.githubusercontent.com/1991226/35512097-614ac626-04f6-11e8-8e83-40a8ea705849.png)

The main differences are:

1. The search bar is optional but, if present, is 100% of its container width and doesn't push the menu bar down.
2. The menu bar is left-aligned in its container
3. The login button, if present, is right-aligned in its container.

Tested in various browsers, including mobile. In some older browsers, the mobile view is used (which is a reasonable failsafe).

Android63 committed:
[refs #00184] Reduce header bottom margin

Should be 15px on desktop.

Android63 committed:
[refs #00184] Add logout button

Add logout button to header in example page.

Android63 committed:
[refs #00184] Left float menu

Float menu bar to the left in its column and login button to the right.

Android63 committed:
[refs #00184] Use correct logo size

Switch to larger logo size, as defined in abstract and sketch.

Android63 committed:
[refs #00184] Add top margin to menu bar

Android63 committed:
[refs #00184] Add top margin to menu button

Android63 committed:
[refs #00184] Change height of search bar

To match height of NHS lozenge.

Android63 committed:
[refs #00184] Resize search icon

To match size of search bar.

Android63 committed:
[refs #00184] Use predefined sizes

Use sizing constants rather than specific dimensions.

Android63 committed:
[refs #00184] Top-align logo and search bar

Using flex box, align header items vertically to top of container.

Android63 committed:
[refs #00184] Use global size definition

...rather than specific value.

Android63 committed:
[refs #00184] adjust margin above menu bar

Adjust top margin for menu bar and login button so text aligns with
first line of logo.

Android63 committed:
[refs #00184] Add negative bottom margin to search

Add negative bottom margin to search bar in header so that, if it is
present, it doesn't push the menu bar down.

Android63 committed:
[refs #00184] Add double-line logo

Add logo with 2 lines of text to test header. Menu bar should stay
vertically aligned with first line of logo text.

Android63 committed:
[refs #00184] Add right-padding to logo

To prevent it running into menu and search bar.

Android63 committed:
[refs #00184] add tall logo

For testing purposes (menu bar should vertically align with first
line of logo text).

cehwitham commented on components/nav-primary/_components.nav-primary.scss:
```diff
@@ -14,8 +14,9 @@
-    float: right;
+    float: left;
     clear: none;
+    margin-top: $global-border-mid+36;
```
Where does the 36 come from here @Android63?

Android63 replied:
It's the height of the search bar (which has a bottom margin of -36px to cancel this out). I've added a comment to that effect.

cehwitham commented on components/search-bar/_components.search-bar.scss:
```diff
@@ -9,13 +9,22 @@
    .c-search__input {
      @extend .c-form-input;
-     height: 38px;
+     height: 36px;
```
Has this always been a fixed height?

Android63 replied:
I think so. The height has to match the NHS lozenge in the logo. I've added a comment to that effect.

cehwitham commented on components/search-bar/_components.search-bar.scss:
```diff
@@ -9,13 +9,22 @@
+   */
+   .c-page-header__controls > .c-search {
+     @include mq($from: md) {
+       margin-bottom: -36px;
```
@csswizardry any thoughts on this? Is this the best way to accomplish this?

csswizardry replied:
This is gonna be an interesting one to solve! Let’s discuss :)

Android63 replied:
Looking forward to it :)

cehwitham commented on index.html:
```diff
@@ -104,7 +104,7 @@
 
       <a href="#0" class="c-page-header__lockup">
-        <img src="./assets/img/logos/nhs-leadership-academy.png" alt="NHS Leadership Academy" width="204" height="62" class="c-page-header__logo" />
+        <img src="./assets/img/logos/NHS_LA_Logo_Blue.png" alt="NHS Leadership Academy" class="c-page-header__logo" />
```
@Android63 should we have a more standardised filenaming convention for images? Seems we're switching between - and _ and caps?

Android63 replied:
Agreed. This is addressed in https://github.com/NHSLeadership/nightingale/pull/179

Android63 committed:
[refs #00184] Add comments

Add comments to clarify search bar height and menu bar top margin.

Android63 commented:
Are you OK with me merging this in @cehwitham? I will ask Harry to review, but I'd like to get the functionality in place for now and correct any issues he raises later.

Android63 committed:
[refs #00184] Utilise BEM mixes in header

Refactor nesting of search in header.

Android63 committed:
[refs #00184] Refactor header's search & nav spacing

Replaced negative margin hack with more elegant solution.

Android63 committed:
[refs #00184] Nest login button inside nav

So login button and nav links can be treated as one component.

csswizardry committed:
[refs #00184] Move font-size: 0; hack

The font-size hack that is used to remove whitespace in the nav was
leaking out into other (namely button) styles. I’ve moved the hack to
sit directly on the nav’s list rather than its wrapping element. This
means that we don’t need to undo its affects on any other elements are
we did previously.

csswizardry committed:
[refs #00184] Remove use of @extends in Search

To reuse the styles of regular form inputs (which is a good thing), we
were using @extends (which is not a good thing). I’ve removed the
@extend directive in favour of using an additional class in the HTML.

csswizardry committed:
[refs #00184] Remove leftover header-button styles

After refactoring the header’s login button, we had some leftover
styles that were no longer used. I’ve removed those now.

csswizardry committed:
[refs #00184] Improve CSS documentation for header

After the header’s refactor, it needed better commenting and
documentation, as well as a general tidying up of formatting.

csswizardry committed:
[refs #00184] Variablise lozenge dimensions

We need to line items in the header up with the NHS lozenge, so we’ve
moved the dimensions out into a global var: this will make updates and
changes much simpler.

csswizardry committed:
[refs #00184] Update clearfix for older browsers

Given that we have quite old browser support, it’s probably an idea to
provide a configurable OldIE version for that eventuality.

csswizardry committed:
[refs #00184] Swap search button for proper sprite

We’ve reworked the main search button to use an actual BUTTON element,
as well as using the spriting element for the icon. Also some nice
sibling selector trickery to do away with the -ve margin that made the
button sit above the search field: they’re now back next to one another
and styled a little more cleanly.

csswizardry committed:
[refs #00184] OldIE fallback for lack of `calc()`

The search input- and button-pair make use of `calc()` to lay themselves
out properly. I’ve built an alternative method for laying things out in
browsers that aren’t as capable.

csswizardry committed:
Merge branch 'develop' into tkt-00184

csswizardry committed:
[refs #00184] Fix header logo

The logo was using the wrong asset URL, plus padding applied to it was
reducing the width of the logo. I’ve corrected the URL and moved the
spacing from the logo to the lockup.

Android63 commented:
@cehwitham I've checked this locally and it looks fine. Could you review it with a view to a merge please?
